From f533bac65fb30d13fc80bac14ab7f589498e64e6 Mon Sep 17 00:00:00 2001
From: Alan Kao <alankao@andestech.com>
Date: Wed, 16 Oct 2019 10:20:47 +0800
Subject: [PATCH 003/191] riscv: andes: Add Andes-specific relocation type

Reformed from the following patches on RISCV-Linux-5.4:
- (d7d722e5f46e) Andes-specific relocation type support
- Reformed from the following patches:
- https://git.andestech.com/alankao/linux/commit/3cb9939499bcf862ed78bfd650701846f2edb966

Signed-off-by: Charles Ci-Jyun Wu <dminus@andestech.com>
---
 arch/riscv/include/uapi/asm/elf.h | 17 ++++++++++++++++
 arch/riscv/kernel/module.c        | 33 ++++++++++++++++++++++++++++++-
 2 files changed, 49 insertions(+), 1 deletion(-)

diff --git a/arch/riscv/include/uapi/asm/elf.h b/arch/riscv/include/uapi/asm/elf.h
index d696d6610231..73f14235e52c 100644
--- a/arch/riscv/include/uapi/asm/elf.h
+++ b/arch/riscv/include/uapi/asm/elf.h
@@ -94,5 +94,22 @@ typedef union __riscv_fp_state elf_fpregset_t;
 #define R_RISCV_SET32		56
 #define R_RISCV_32_PCREL	57
 
+/* Andes V5 */
+#define R_RISCV_ALIGN_BTB	240
+#define R_RISCV_10_PCREL	241
+#define R_RISCV_DATA		242
+#define R_RISCV_LALO_HI20	243
+#define R_RISCV_LALO_LO12_I	244
+#define R_RISCV_RELAX_ENTRY	245
+#define R_RISCV_LGP18S0		246
+#define R_RISCV_LGP17S1		247
+#define R_RISCV_LGP17S2		248
+#define R_RISCV_LGP17S3		249
+#define R_RISCV_SGP18S0		250
+#define R_RISCV_SGP17S1		251
+#define R_RISCV_SGP17S2		252
+#define R_RISCV_SGP17S3		253
+#define R_RISCV_RELAX_REGION_BEGIN	254
+#define R_RISCV_RELAX_REGION_END	255
 
 #endif /* _UAPI_ASM_RISCV_ELF_H */
diff --git a/arch/riscv/kernel/module.c b/arch/riscv/kernel/module.c
index 91fe16bfaa07..1f7be2651b29 100644
--- a/arch/riscv/kernel/module.c
+++ b/arch/riscv/kernel/module.c
@@ -296,6 +296,25 @@ static int apply_r_riscv_sub64_rela(struct module *me, u32 *location,
 	return 0;
 }
 
+static int apply_r_riscv_10_pcrel_rela(struct module *me, u32 *location,
+				       Elf_Addr v)
+{
+	s64 offset = (void *)v - (void *)location;
+	u32 imm10 = (offset & 0x400) << (31 - 10);
+	u32 imm9_5 = (offset & 0x3e0) << (25 - 5);
+	u32 imm4_1 = (offset & 0x1e) << (8 - 1);
+
+	*(u32 *) location = (*(u32 *) location & 0x41fff0ff) |
+	    imm10 | imm9_5 | imm4_1;
+	return 0;
+}
+
+static int apply_r_riscv_ignore_rela(struct module *me, u32 *location,
+				     Elf_Addr v)
+{
+	return 0;
+}
+
 static int (*reloc_handlers_rela[]) (struct module *me, u32 *location,
 				Elf_Addr v) = {
 	[R_RISCV_32]			= apply_r_riscv_32_rela,
@@ -321,6 +340,18 @@ static int (*reloc_handlers_rela[]) (struct module *me, u32 *location,
 	[R_RISCV_SUB64]			= apply_r_riscv_sub64_rela,
 };
 
+static int (*reloc_handlers_rela_nds(unsigned int type)) (struct module *me,
+							  u32 *location,
+							  Elf_Addr v)
+{
+	if (type == R_RISCV_10_PCREL)
+		return apply_r_riscv_10_pcrel_rela;
+	else if (type >= R_RISCV_ALIGN_BTB && type <= R_RISCV_RELAX_REGION_END)
+		return apply_r_riscv_ignore_rela;
+	else
+		return NULL;
+}
+
 int apply_relocate_add(Elf_Shdr *sechdrs, const char *strtab,
 		       unsigned int symindex, unsigned int relsec,
 		       struct module *me)
@@ -357,7 +388,7 @@ int apply_relocate_add(Elf_Shdr *sechdrs, const char *strtab,
 		if (type < ARRAY_SIZE(reloc_handlers_rela))
 			handler = reloc_handlers_rela[type];
 		else
-			handler = NULL;
+			handler = reloc_handlers_rela_nds(type);
 
 		if (!handler) {
 			pr_err("%s: Unknown relocation type %u\n",
-- 
2.34.1

